fixes DOCS-159 (#248)
pinot-contrib/pinot-docs: developers/advanced/advanced-pinot-setup.md

### Add an inverted index to your table automatically

It is possible to force the creation of an inverted index automatically by adding an entry to your [table index config](../../configuration-reference/table.md#table-index-config) in the table configuration file. This works whether you are creating a [batch (offline) table](#batch-table-creation) or a [streaming (real-time) table](#streaming-table-creation) and is accomplished by setting the `createInvertedIndexDuringSegmentGeneration` flag to `true` in your table config, as follows:

```json
...
"tableIndexConfig": {
    ...
    "createInvertedIndexDuringSegmentGeneration": true,
    ...
}
...
```

During the segment generation process, Pinot checks to see if this is `true`. The setting is `false` by default. If true, Pinot creates an inverted index for the columns that you specify in the `invertedIndexColumns` list in the table configuration. If false, no inverted index is created.

If you update this setting in your table configuration, you must [reload the table segment](../../basics/data-import/segment-reload.md) to apply the inverted incex to all existing segments.

> You cannot apply an inverted index to a consuming or non-consuming segment.

Author: I can't find a definition of either consuming or non-consuming segments to link to or write here.

Author: I found and added one

See [`examples`](https://github.com/apache/pinot/tree/master/pinot-tools/src/main/resources/examples) for all possible batch/streaming tables.

### Add an inverted index to your table automatically

> It is possible to force the creation of an inverted index automatically by adding an entry to your [table index config](../../configuration-reference/table.md#table-index-config) in the table configuration file. This works whether you are creating a [batch (offline) table](#batch-table-creation) or a [streaming (real-time) table](#streaming-table-creation) and is accomplished by setting the `createInvertedIndexDuringSegmentGeneration` flag to `true` in your table config, as follows:

Reviewer: Actually this config is only applicable to batch tables, in the basic setup. In streaming tables, everything is automatically created. Unless you're manually pushing to the streaming table, but that's a special operation anyway

```json
...
"tableIndexConfig": {
    ...
    "createInvertedIndexDuringSegmentGeneration": true,
    ...
}
...
```

During the segment generation process, Pinot checks to see if this is `true`. The setting is `false` by default. If true, Pinot creates an inverted index for the columns that you specify in the `invertedIndexColumns` list in the table configuration. If false, no inverted index is created.

If you update this setting in your table configuration, you must [reload the table segment](../../basics/data-import/segment-reload.md) to apply the inverted index to all existing segments.

> You cannot apply an inverted index to a consuming or non-consuming [segment](../../basics/components/table/segment/).

Reviewer: What are you referring to here, when you say non-consuming segment?

Reviewer: Consuming segments also can have inverted index. Just this config is not relevant to them

Author: > What are you referring to here, when you say non-consuming segment?

you know what? I think I got that from an AI response when I was learning about segments and failed to do better research. I can't find that such a thing exists. Thank you for calling that out, @npawar !

{% endtab %}
{% endtabs %}

### Automatically add an inverted index to your batch table

> It is possible to force the creation of an inverted index during segment generation by adding an entry to your [table index config](../../configuration-reference/table.md#table-index-config) in the table configuration file.

Reviewer 2: ```suggestion
For [batch (offline) tables](#batch-table-creation), you can forcibly create an inverted index during segment generation by setting `createInvertedIndexDuringSegmentGeneration` to `true` in in the table configuration file in the [table index config](../../configuration-reference/table.md#table-index-config) section. The following example shows how this configuration looks:
```

Reviewer 2: what is the alternative to not forcing this? may be helpful to include when this happens naturally (assuming there's another way, being we're using the term force)

> This setting works with [batch (offline) tables](#batch-table-creation). This setting is `false` by default but you can set `createInvertedIndexDuringSegmentGeneration` to `true` in your table config, as follows:

Reviewer 2: ```suggestion
```

```json
...
"tableIndexConfig": {
    ...
    "createInvertedIndexDuringSegmentGeneration": true,
    ...
}
...
```

> When set to `true`, Pinot creates an inverted index for the columns that you specify in the `invertedIndexColumns` list in the table configuration. If `false`, no inverted index is created.

Reviewer 2: is this still only for batch tables? if yes, we may want to combine with the first paragraph as an overview for what the whole thing is, give the example, then say, what you must do after to apply the index...

When you update this setting in your table configuration, you must [reload the table segment](../../basics/data-import/segment-reload.md) to apply the inverted index to all existing segments.

{% hint style="info" %}

> All other index types are created during the segment generation by default, and that is the preferred thing to do. However, since this is not the default for inverted indexes, this setting enables you to also create them at segment generation time.

Reviewer 2: it sounds like we may be able to drop the word force....by default, the inverted index type is the only type of index that isn't created automatically during segment generation. To automatically create ....set to true

Waiting to build indexes until load time increases the startup time and takes up resources with every new segment push, which increases the time for other operations such as rebalance.

> {% endhint %}

Reviewer 2: ah K, think this answer my question above about diff between creating an index when the segment is first created or waiting until it's uploaded to the server. May help to move/consolidate this up with the first paragraph, explaining the whole concept

{% endtab %}
{% endtabs %}

### Use `sortedColumn` with streaming tables

> For [streaming](#streaming-table-creation) tables, you can use a sorted index with `sortedColumn`. This will sort data when generating segments as the segment is committed. See [Real-time tables](../../basics/indexing/forward-index.md#real-time-tables) for more information.

Reviewer 2: ```suggestion
For [streaming](#streaming-table-creation) tables, you can use a sorted index with `sortedColumn` to sort data when generating segments as the segment is committed. See [Real-time tables](../../basics/indexing/forward-index.md#real-time-tables) for more information.
```

Reviewer 2: does committed here mean the same as created? if so, perhaps we say that, if not, lets clarify the distinction
